workers/unity/Packages/io.improbable.gdk.tools/.CodeGenDocs/2-write-a-generator.md

```diff
@@ -0,0 +1,260 @@
+# Write a generator with the CodeWriter API
+
+### _"It's lambdas all the way down!"_
+
+## Set up a generator
+
+Create a static class in your code generator's namespace that implements a `Generate` method, returning a `CodeWriter`.
+
+For example:
+
+```csharp
+using Improbable.Gdk.CodeGeneration.CodeWriter;
+using Improbable.Gdk.CodeGeneration.Model.Details;
+using NLog;
+
+namespace Your.Own.CodeGenerator
+{
+    public static class YourFirstGenerator
+    {
+        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+
+        public static CodeWriter Generate()
+        {
+            // Create and return a CodeWriter
+        }
+    }
+}
+```
+
+> Note: `NLog` is optional, but recommended for code generator logging.
+
+### Making use of `AddGenerators`
+
+To take advantage of the `AddGenerators` method on the `CodegenJob`, your function must match one of the following signatures:
+
+```csharp
+public static CodeWriter Generate(UnityComponentDetails details);
+public static CodeWriter Generate(UnityTypeDetails details);
+public static CodeWriter Generate(UnityEnumDetails details);
+```
+
+> Note: the methods do not _have_ to be called `Generate` nor be public. Their accessibility should depend on where you are calling them from.
```

Comment:
but the return value and the arguments need to be matching, I assume? 

I wonder if there is a better way to phrase that though this is definitely clearer than it was before

Reply:
```suggestion
> Note: the methods do not _have_ to be called `Generate`, but they must return a `CodeWriter` and consume a type of `GeneratorInputDetails`. They also do not have to be public: accessibility should depend on where you are calling them from.
```
?